File: src/main/java/com/github/puregero/warpplugin/WarpCommand.java
package com.github.puregero.warpplugin;

import net.kyori.adventure.text.Component;
import net.kyori.adventure.text.format.NamedTextColor;
import org.bukkit.Location;
import org.bukkit.command.Command;
import org.bukkit.command.CommandExecutor;
import org.bukkit.command.CommandSender;
import org.bukkit.command.TabCompleter;
import org.bukkit.entity.Player;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.List;
import java.util.Objects;

public class WarpCommand implements CommandExecutor, TabCompleter {
    private final WarpPluginPlugin plugin;

    public WarpCommand(WarpPluginPlugin plugin) {
        this.plugin = plugin;
        Objects.requireNonNull(plugin.getCommand("warp")).setExecutor(this);
        Objects.requireNonNull(plugin.getCommand("warp")).setTabCompleter(this);
    }

    @Override
    public boolean onCommand(@NotNull CommandSender sender, @NotNull Command command, @NotNull String label, @NotNull String[] args) {
        if (args.length == 0) {
            sender.sendMessage(Component.text("Usage: /%s <warp>".formatted(label)).color(NamedTextColor.RED));
            return false;
        }

        String warpName = args[0];
        Location warpLocation = this.plugin.getWarpLocation(warpName);

        if (warpLocation == null) {
            sender.sendMessage(Component.text("Warp not found: " + warpName).color(NamedTextColor.RED));
            return false;
        }

        if (!(sender instanceof Player player)) {
            sender.sendMessage(Component.text("Only players can execute this command!").color(NamedTextColor.RED));
            return false;
        }

        player.sendMessage(Component.text("Teleporting...").color(NamedTextColor.GREEN));
        player.teleportAsync(warpLocation).thenAccept(teleported ->
            player.sendMessage(Component.text("Teleported to warp: " + warpName).color(NamedTextColor.GREEN))
        );

        return true;
    }

    @Override
    public @Nullable List<String> onTabComplete(@NotNull CommandSender commandSender, @NotNull Command command, @NotNull String s, @NotNull String[] args) {
        if (args.length == 1) {
            return this.plugin.getWarpNames().stream().filter(warpName -> warpName.toLowerCase().startsWith(args[0].toLowerCase())).toList();
        }
        return List.of();
    }
}

File: src/main/java/com/jjkay03/foliawarp/commands/WarpCommand.kt
package com.jjkay03.foliawarp.commands

import com.jjkay03.foliawarp.FoliaWarp
import org.bukkit.Bukkit
import org.bukkit.command.Command
import org.bukkit.command.CommandExecutor
import org.bukkit.command.CommandSender
import org.bukkit.command.TabCompleter
import org.bukkit.entity.Player

class WarpCommand : CommandExecutor, TabCompleter {

    init {
        FoliaWarp.INSTANCE.getCommand("warp")?.setExecutor(this)
        FoliaWarp.INSTANCE.getCommand("warp")?.tabCompleter = this
    }

    // COMMAND
    override fun onCommand(sender: CommandSender, command: Command, label: String, args: Array<out String>): Boolean {
        // End if invalid arg
        if (args.isEmpty()) { sender.sendMessage("§cUsage: /$label <warp> [player]"); return false }

        val warpName = args[0]
        val warpLocation = FoliaWarp.INSTANCE.getWarpLocation(warpName)

        // End if warp not found
        if (warpLocation == null) { sender.sendMessage("§cWarp not found!"); return false }

        // Check if player argument provided
        if (args.size >= 2) {
            val targetPlayerName = args[1]
            val targetPlayer = Bukkit.getPlayer(targetPlayerName)

            // End if target player not found
            if (targetPlayer == null) { sender.sendMessage("§cPlayer not found!"); return false }

            // Teleport target player
            targetPlayer.teleportAsync(warpLocation).thenAccept { _ ->
                targetPlayer.sendMessage("§3\uD83C\uDF00 §bYou have been warped to §3$warpName")
                sender.sendMessage("§3\uD83C\uDF00 §bWarped §r${targetPlayer.name}§b to §3$warpName")
            }
        } else {
            // End if not player
            if (sender !is Player) { sender.sendMessage("§cOnly players can execute this command!"); return false }

            // Teleport sender
            sender.teleportAsync(warpLocation).thenAccept { _ ->
                sender.sendMessage("§3\uD83C\uDF00 §bWarped to §3$warpName")
            }
        }

        return true
    }

    // TAB COMPLETION
    override fun onTabComplete(sender: CommandSender, command: Command, label: String, args: Array<out String>): List<String>? {
        if (args.size == 1) {
            return FoliaWarp.INSTANCE.getWarpNames().filter { it.lowercase().startsWith(args[0].lowercase()) }
        }
        if (args.size == 2) {
            return Bukkit.getOnlinePlayers().map { it.name }.filter { it.lowercase().startsWith(args[1].lowercase()) }
        }
        return emptyList()
    }

}
